src/v/cluster/rm_stm.cc

```diff
@@ -1362,6 +1362,14 @@ ss::future<result<kafka_result>> rm_stm::replicate_seq(
         co_return errc::invalid_request;
     }
 
+    ss::lw_shared_ptr<mutex> idempotent_lock = nullptr;
+    std::optional<ss::semaphore_units<ss::named_semaphore_exception_factory>>
+      units;
+    if (!bid.is_transactional) {
+        idempotent_lock = get_idempotent_producer_lock(bid.pid);
+        units = co_await idempotent_lock->get_units();
```

Comment:
we need to take read lock because `replicate_seq` can process several requests issued concurrently by the same pid; if we use regular log we'll kill performance